### PR TITLE
Progress summary RAPTURE-38

### PR DESCRIPTION
**Pull request by macklin-10x**
Augments the rclone RemoteStats method to include a couple of pieces of data necessary for computing overall transfer progress.

### CODE REVIEW

**Comment by andreasg123** [comment]
I'm not happy with the output formatting on the server side.  It looks like you modeled it after `func (s *StatsInfo) String()`.  I would really prefer something modeled after `func (s *StatsInfo) RemoteStats()`, maybe even just converting `rc.Params` to JSON.  I don't know if the detailed information on every file being transferred is needed but it might be nice in a clickable UI that can reveal more detail.  In any case, I would like the data to be transferred to the client and formatted there.

If you disagree with this high-level assessment, I can instead review the PR as is.

**Comment by macklin-10x** [comment]
I agree with you in spirit, but I'd prefer to leave this as-is.  `RemoteStats` doesn't include all of the data needed to reconstruct the full summary returned from `String`, and since the rapture agent "server" serves exactly one client - the frontend via localhost - the API boundary doesn't need to be that strict.  Nothing besides the client on localhost is going to access this API.  I'd prefer to just lean on the existing rclone string formatting code rather than re-implementing it in javascript.

The main things missing from `RemoteStats` are the overall transfer-level statistics that tell the user the most important thing - "when will be entire upload be complete?".

Please review as-is for now so I can get the feature released, and I'll consider coming back in a follow-up to make the handler return something more data-centric rather than preformatted strings.

**Comment by macklin-10x** [comment]
Hmm that said I could always add a method that augments the data returned from `RemoteStats` with the few missing fields.  That might be nicer.  I'll take a stab at that tomorrow morning, please hold off for now on review.

**Comment by andreasg123** [comment]
I hope that I didn't come across as being overly critical.  I'm still feeling my way into a proper code review etiquette because my colleagues in my last job couldn't be bothered with doing code reviews.

I made my suggestion for several reasons.  First, there is the separation of concerns.  In my opinion, the backend should provide data but perform no formatting with the exception of SSR.  Second, there is internationalization.  I actually haven't looked up what our colleagues in Sweden would prefer but Germany uses a decimal comma instead of a decimal point.  Third, as I mentioned earlier, information about individual file transfers might be a nice addition to the UI by letting the user click on something.

Regarding the overall transfer-level statistics, that could be recomputed from the data returned by `RemoteStats`.  However, I'm not completely sure if replicating that logic at the client is really the right approach.

**Comment by andreasg123** [comment]
I saw somewhere a desire for a UI to see what all Rapture clients are doing (can't find it now).  This could be a starting point for such a global dashboard.

**Comment by macklin-10x** [comment]
"Regarding the overall transfer-level statistics, that could be recomputed from the data returned by `RemoteStats`"

There actually isn't enough info in the current RemoteStats method to do this; there are a few necessary fields missing.  That said, I think augmenting RemoteStats to include them is the way to go.

**Comment by macklin-10x** [comment]
@andreasg123 ready for review

File: fs/accounting/stats.go [hunk 1]
```diff
@@ -55,10 +55,19 @@ func NewStats() *StatsInfo {
 
 // RemoteStats returns stats for rc
 func (s *StatsInfo) RemoteStats() (out rc.Params, err error) {
+	// checking and transferring have their own locking so read
+	// here before lock to prevent deadlock on GetBytes
+	transferring, checking := s.transferring.count(), s.checking.count()
+	transferringBytesDone, transferringBytesTotal := s.transferring.progress(s)
```
Comment (andreasg123) [inline fs/accounting/stats.go:61]: I don't know if it matters but this acquires `s.transferring.my.RLock()` twice plus one more time each for `s.transferring` and `s.checking` in the loops below. If acquiring the lock is expensive, it should be combined.

File: fs/accounting/stats.go [hunk 2]
```diff
@@ -55,10 +55,19 @@ func NewStats() *StatsInfo {
...
+
 	out = make(rc.Params)
 	s.mu.RLock()
 	out["speed"] = s.Speed()
 	out["bytes"] = s.bytes
+	// note that s.bytes already includes transferringBytesDone so
+	// we take it off here to avoid double counting
+	// This is added for Rapture for computing overall transfer ETA.
+	out["size"] = s.transferQueueSize + s.bytes + transferringBytesTotal - transferringBytesDone
```
Comment (andreasg123) [inline fs/accounting/stats.go:70]: It's a good idea just to compute those properties here instead of having to do it on the JavaScript side.

**Review by andreasg123 — APPROVED** [review]
Looks good

File: fs/accounting/stats.go (hunk 1 above)
Comment (macklin-10x) [reply fs/accounting/stats.go:61]: Locks in go are very lightweight due to the use of green threads.  Note that the existing implementation of `String` does the same thing (I lifted this implementation from there).

**Pull request merged**